### PR TITLE
Updating the original files in libraries to Dafny 4.

### PR DESCRIPTION
**Pull request by davidcok**
By submitting this pull request, I confirm that my contribution is made under the terms of the MIT license.

### CODE REVIEW

**Comment by robin-aws** [comment]
Nit: Mention in the PR body/commit message that you're also adding the autoformatter. I was about to ask if you got the PR title wrong because there are so few changes for Dafny 4.0 compatibility, I wasn't even sure that part was happening. :D

File: pydiff.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
+#!/usr/bin/env python
```
Comment (robin-aws) [inline pydiff.py:1]: I don't LOVE adding this file and digging ourselves even deeper into `lit`, but I'm OKAY with it for now. Once we have project files we should definitely tackle #35 though.

**Review by robin-aws — CHANGES_REQUESTED** [review]
Looks great (and I mean that more literally than usual, thank you @MikaelMayer's autoformatter! :)

My only ask is if we're claiming this makes it Dafny 4.0 compatible, we have to back that up in CI. Since we can't just ask `setup-dafny-action` to download 4.0 yet, could we set up lit to support substituting `dafny --functionSyntax:4 --quantifierSyntax:4 --unicode-char ...` as `%dafny` instead? Then we can add a build matrix to the CI to run both with 3.x dafny and with 3.x dafny with those flags.

File: pydiff.py (hunk 1 above)
Comment (davidcok) [reply pydiff.py:1]: If we make nightly builds right after 3.12 that have 4.0 defaults, then we can just list a bunch of specific builds to test against. I have tested them locally on my machine, running lit with a 4.0 build.

**Comment by davidcok** [comment]
Using lit made it very easy to setup and update the testing though.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # This workflow performs static analysis checks.
-name: tests
+name: Legacy verification
```
Comment (robin-aws) [inline .github/workflows/tests.yml:2]: Revert since this doesn't make sense without other edits.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -17,23 +17,24 @@ jobs:
       - name: Install Dafny
         uses: dafny-lang/setup-dafny-action@v1
         with:
-          dafny-version: "3.9.1"
+          dafny-version: "nightly-2023-02-15-567a5ba"
```
Comment (robin-aws) [inline .github/workflows/tests.yml:20]: We shouldn't have the main CI using a nightly build

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -17,23 +17,24 @@ jobs:
...
-
-      - uses: actions/upload-artifact@v2  # upload test results
-        if: success() || failure()        # run this step even if previous step failed
-        with:
-          name: verification-results
-          path: '**/TestResults/*.trx'
-          if-no-files-found: error
+      - name: Verify Legacy Code
+        run: lit --time-tests -v 
+        ##--param 'dafny_params=/verificationLogger:trx /verificationLogger:csv' .
```
Comment (robin-aws) [inline .github/workflows/tests.yml:30]: We shouldn't be disabling parts of the CI, even temporarily.

**Comment by robin-aws** [comment]
Dismissed my earlier Request Changes as I'm okay with not testing in Dafny 4.0 mode for now, since it will be easier to do that once 4.0 is released. I still don't want the CI to be put in a broken state temporarily but I will let other reviewers help address that.

File: .github/workflows/tests.yml (hunk 4 above)
Comment (atomb) [reply .github/workflows/tests.yml:30]: It looks like, in particular, the missing `.` at the end is causing the tests to fail at the moment.

**Comment by atomb** [comment]
I think the current test failures may be due to changing the path to be verified from `.` to `src`, since that may have led to the test results being stored in a different directory. I think it makes sense to include the examples in verification, too, which would suggest leaving the target directory as `.`.

**Comment by davidcok** [comment]
Agree that the path should be ., not src. Thanks for spotting that.

**Comment by atomb** [comment]
Hmm. Still failing. If you run the `lit` command locally, do you get any `.trx` files generated?

**Comment by davidcok** [comment]
PR shifted to #97 for a continuing branch

**Pull request closed without merging**